src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,182 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int currentSize;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (currentSize == 0) {
+            addFirstNode(value);
+            return;
+        }
+        addNodeToTailPosition(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > currentSize) {
+            throw new IndexOutOfBoundsException("Input index : " + index
+                    + " out of bonds: [" + 0 + ", " + currentSize + "]");
+        } else if (currentSize == index) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            addNodeToHeadPosition(value);
+            return;
+        }
+        addNodeToAnyPosition(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexBounds(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexBounds(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T temp = currentNode.value;
+        currentNode.value = value;
+        return temp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexBounds(index);
+        if (index == 0 && currentSize == 1) {
+            T tempToReturn = head.value;
+            unlinkSingleNode();
+            return tempToReturn;
+        }
+        Node<T> currentNode = getNodeByIndex(index);
+        remove(currentNode.value);
+        return currentNode.value;
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public boolean remove(T value) {
+        if (isValuesEquals(head.value, value) && currentSize == 1) {
+            return unlinkSingleNode();
+        } else if (isValuesEquals(head.value, value)) {
+            return unlinkHeadNode();
+        }
```

Comment:
I don't think we need an exclusive check of the head, I'd rather move the part of handling the case of removing the last element to the `unlink` method, to have a single one of these, and passed to it not a value but a node. 
Be careful of redundant iterations to find the node you are trying to delete

Reply:
I don't create one method unlink, where I can check Node position and unlink node from List, because I think that the single-responsibility method for unlink or add Node is more readable and simple for understanding, to the other dev which read my code first time and for me after some time. If this approach to writing code incorrect i can rewrite this logic.

Reply:
I think it is better to listen to Vadim's advice, because all three methods perform the same function depending on the position of the element.